File: app/src/main/java/com/wavesplatform/wallet/payload/OrderBook.java
package com.wavesplatform.wallet.payload;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Created by anonymous on 01.07.17.
 */

public class OrderBook {

    @SerializedName("timestamp")
    public String timestamp;
    @SerializedName("pair")
    public Pair pair;
    @SerializedName("bids")
    public List<Bids> bids;
    @SerializedName("asks")
    public List<Asks> asks;

    public static class Pair {
        @SerializedName("amountAsset")
        public String amountAsset;
        @SerializedName("priceAsset")
        public String priceAsset;
    }

    public static class Bids implements Parcelable{
        @SerializedName("price")
        public long price;
        @SerializedName("amount")
        public String amount;

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Bids bids = (Bids) o;

            if (price != bids.price) return false;
            return amount != null ? amount.equals(bids.amount) : bids.amount == null;

        }

        @Override
        public int hashCode() {
            int result = (int) (price ^ (price >>> 32));
            result = 31 * result + (amount != null ? amount.hashCode() : 0);
            return result;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeLong(this.price);
            dest.writeString(this.amount);
        }

        public Bids() {
        }

        protected Bids(Parcel in) {
            this.price = in.readLong();
            this.amount = in.readString();
        }

        public static final Creator<Bids> CREATOR = new Creator<Bids>() {
            @Override
            public Bids createFromParcel(Parcel source) {
                return new Bids(source);
            }

            @Override
            public Bids[] newArray(int size) {
                return new Bids[size];
            }
        };
    }

    public static class Asks implements Parcelable{
        @SerializedName("price")
        public long price;
        @SerializedName("amount")
        public String amount;

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Asks asks = (Asks) o;

            if (price != asks.price) return false;
            return amount != null ? amount.equals(asks.amount) : asks.amount == null;

        }

        @Override
        public int hashCode() {
            int result = (int) (price ^ (price >>> 32));
            result = 31 * result + (amount != null ? amount.hashCode() : 0);
            return result;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeLong(this.price);
            dest.writeString(this.amount);
        }

        public Asks() {
        }

        protected Asks(Parcel in) {
            this.price = in.readLong();
            this.amount = in.readString();
        }

        public static final Creator<Asks> CREATOR = new Creator<Asks>() {
            @Override
            public Asks createFromParcel(Parcel source) {
                return new Asks(source);
            }

            @Override
            public Asks[] newArray(int size) {
                return new Asks[size];
            }
        };
    }
}

File: app/src/main/java/com/wavesplatform/wallet/v2/data/model/remote/response/OrderBook.kt
/*
 * Created by Eduard Zaydel on 1/4/2019
 * Copyright © 2019 Waves Platform. All rights reserved.
 */

package com.wavesplatform.wallet.v2.data.model.remote.response

import android.util.Log
import com.chad.library.adapter.base.entity.MultiItemEntity
import com.google.common.primitives.Bytes
import com.google.gson.annotations.SerializedName
import com.wavesplatform.wallet.v1.util.MoneyUtil
import com.wavesplatform.wallet.v1.util.SignUtil
import com.wavesplatform.wallet.v2.ui.home.dex.trade.orderbook.TradeOrderBookAdapter
import com.wavesplatform.wallet.v2.util.stripZeros
import java.math.BigInteger

data class OrderBook(
        @SerializedName("timestamp") var timestamp: Long = 0,
        @SerializedName("pair") var pair: Pair = Pair(),
        @SerializedName("bids") var bids: List<Bid> = listOf(),
        @SerializedName("asks") var asks: List<Ask> = listOf()
) {

    data class Pair(
            @SerializedName("amountAsset") var amountAsset: String? = "",
            @SerializedName("priceAsset") var priceAsset: String? = ""
    ) {
        fun toBytes(): ByteArray {
            return try {
                Bytes.concat(SignUtil.arrayOption(amountAsset),
                        SignUtil.arrayOption(priceAsset))
            } catch (e: Exception) {
                Log.e("Wallet", "Couldn't create bytes for AssetPair: ", e)
                ByteArray(0)
            }
        }
    }

    data class Ask(
            @SerializedName("amount") var amount: Long = 0,
            @SerializedName("price") var price: Long = 0,
            @SerializedName("sum") var sum: Long = 0
    ) : MultiItemEntity {
        override fun getItemType(): Int {
            return TradeOrderBookAdapter.ASK_TYPE
        }

        val total: Long
            get() {
                return BigInteger.valueOf(amount)
                        .multiply(BigInteger.valueOf(price))
                        .divide(BigInteger.valueOf(100000000)).toLong()
            }

        fun getScaledSum(priceAssetDecimals: Int?): String {
            return MoneyUtil.getTextStripZeros(sum, priceAssetDecimals ?: 8).stripZeros()
        }
    }

    data class Bid(
            @SerializedName("amount") var amount: Long = 0,
            @SerializedName("price") var price: Long = 0,
            @SerializedName("sum") var sum: Long = 0
    ) : MultiItemEntity {
        override fun getItemType(): Int {
            return TradeOrderBookAdapter.BID_TYPE
        }

        val total: Long
            get() {
                return BigInteger.valueOf(amount)
                        .multiply(BigInteger.valueOf(price))
                        .divide(BigInteger.valueOf(100000000)).toLong()
            }


        fun getScaledSum(priceAssetDecimals: Int?): String {
            return MoneyUtil.getTextStripZeros(sum, priceAssetDecimals ?: 8).stripZeros()
        }
    }
}